node/actors/bft/src/testonly/twins/scenario.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::BTreeSet;
+
+use rand::{seq::SliceRandom, Rng};
+
+use super::{splits, HasKey, Split, Twin};
+
+/// A cluster holds all the nodes in the simulation, some of which are twins.
+pub struct Cluster<T> {
+    /// All nodes, twins and originals.
+    nodes: Vec<T>,
+    /// The ideal size of the committee, ie. the `n` in the `n = 5 * f + 1`.
+    num_replicas: usize,
+}
+
+impl<T> Cluster<T>
+where
+    T: Twin,
+{
+    /// Take a number of non-twin replicas and produce a cluster with a desired number of twins.
+    pub fn new(replicas: Vec<T>, num_twins: usize) -> Self {
+        // Check that the replicas aren't twins.
+        assert_eq!(
+            unique_key_count(&replicas),
+            replicas.len(),
+            "replicas must have unique keys"
+        );
+
+        let twins = replicas
+            .iter()
+            .take(num_twins)
+            .map(|r| r.to_twin())
+            .collect::<Vec<_>>();
+
+        let num_replicas = replicas.len();
+
+        let mut nodes = replicas;
+        nodes.extend(twins);
+
+        Self {
+            nodes,
+            num_replicas,
+        }
+    }
+
+    /// The number of nodes in the cluster, both replicas and twins.
+    pub fn num_nodes(&self) -> usize {
+        self.nodes.len()
+    }
+
+    /// The number of original replicas, ie. not twins.
+    ///
+    /// This is the `n` in the `n = 5 * f + 1` equation.
+    pub fn num_replicas(&self) -> usize {
+        self.num_replicas
+    }
+
+    /// The number of twins in the cluster.
+    pub fn num_twins(&self) -> usize {
+        self.num_nodes() - self.num_replicas()
+    }
+
+    /// The number of faulty nodes the committee can tolerate for the `n = 5 * f + 1` relation to hold.
+    pub fn num_faulty(&self) -> usize {
+        (self.num_replicas - 1) / 5
+    }
+
+    /// The number of signatures required for a quorum, ie. `4 * f + 1` or more generally `n - f`, which in this case equals `n*4/5+1`
+    pub fn quorum_size(&self) -> usize {
+        self.num_replicas() - self.num_faulty()
+    }
+
+    /// The number of signatures required for a high-vote, ie. `2 * f + 1`
+    pub fn subqourum_size(&self) -> usize {
+        self.num_faulty() * 2 + 1
```

Comment:
This should be `self.num_fault() * 2 / 5 + 1` for it not to degenerate into 1 if `num_replicas <= 5`.

Reply:
this is not a degenerate case. If you want to change it, you need to prove correctness of the algorithm for all possible values of n.

Reply:
So we start with `n = 5f+1` and therefore `f = (n-1)/5`. @brunoffranca says the correct formula would be `sq = n-3*f`, so let's see: 

```
sq = n-3*f 
sq = n - 3 * (n-1)/5
sq = n - 3/5n + 3/5
sq = 2/5n + 3/5
```

Now let's start from the current `sq = 2f+1`:
```
sq = 2f+1
sq = 2*(n-1)/5 + 1
sq = 2/5*n -2/5 + 1
sq = 2/5n + 3/5
```

I believe these are equivalent in mathematical terms, however due to the integer arithmetic, for n<6 we have f=0, but that's just an artifact of how computers work. 

Similarly how the spec says the commit quorum size is 4f+1, when n=5 it gives 1, which is clearly wrong, for the same reason I think it's wrong for 2f+1, except we use it for the more difficult to understand subquorum, where nobody who sees it first is really sure if it's an optimisation or a required for safety. 

Here are the values for the different values of `n`:
```python
>>> for t in [(n, f, 2*f+1, 2*n//5+1, n-3*f) for (n, f) in  [(n, (n-1)//5) for n in range(1, 15)]]: print("n={} f={} sq_2f1={} sq_25n1={} sq_n3f={}".format(*t))
... 
n=1 f=0 sq_2f1=1 sq_25n1=1 sq_n3f=1
n=2 f=0 sq_2f1=1 sq_25n1=1 sq_n3f=2
n=3 f=0 sq_2f1=1 sq_25n1=2 sq_n3f=3
n=4 f=0 sq_2f1=1 sq_25n1=2 sq_n3f=4
n=5 f=0 sq_2f1=1 sq_25n1=3 sq_n3f=5
n=6 f=1 sq_2f1=3 sq_25n1=3 sq_n3f=3
n=7 f=1 sq_2f1=3 sq_25n1=3 sq_n3f=4
n=8 f=1 sq_2f1=3 sq_25n1=4 sq_n3f=5
n=9 f=1 sq_2f1=3 sq_25n1=4 sq_n3f=6
n=10 f=1 sq_2f1=3 sq_25n1=5 sq_n3f=7
n=11 f=2 sq_2f1=5 sq_25n1=5 sq_n3f=5
n=12 f=2 sq_2f1=5 sq_25n1=5 sq_n3f=6
n=13 f=2 sq_2f1=5 sq_25n1=6 sq_n3f=7
n=14 f=2 sq_2f1=5 sq_25n1=6 sq_n3f=8
```

I think if `n-3f` is correct and `2f+1` is correct then the middle ground must be correct too. You can see they don't give the exact same results all across the board, but of course the starkest difference is below n=6. 

If it's a safety feature, ie. if there is exactly one highest subquorum then you _must_ repropose it, then I'm a bit worried that if the sq=1, it's more likely that we _don't_ repropose because we see two of them, and lose this security. I don't have an example for it from the top of my head, but it seems more intuitive for the other values of 2f+1 that if there are two of them then there surely wasn't any quorum. For example if n=5 and a leader go isolated and voted for their own proposal, but missed the fact that there was a quorum for a value already in the previous round.